{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 50\n",
    "fastp = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ShigaTyper Shigella serotype prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "now = datetime.datetime.now().strftime(\"%Y%m%d\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load samples for analysis\n",
    "# identify sample names for the pair-end reads\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd \n",
    "files = glob.glob(\"*.fastq.gz\")\n",
    "names = []\n",
    "for file in files:\n",
    "    names.append(file[:file.find(\"_\")])\n",
    "samples = list(set(names))\n",
    "samples.sort()\n",
    "sizes = []\n",
    "for Sample in samples:\n",
    "    size = 0\n",
    "    for file in files:\n",
    "        if file.startswith(Sample+\"_\"):\n",
    "            size += os.path.getsize(file)\n",
    "    sizes.append(round(size/1024000, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you need to change the directory/file for the Shigella reference sequence database:\n",
    "# Remove \"#\" from the beginning of the 3 lines below starting from \"import os\" \n",
    "# Replace the absolute directory and file name within quotation marks for the reference sequence database \n",
    "# in the line beginning with \"AbsRef_dir\" after \"=\" \n",
    "# For example, '/mnt/hgfs/' is the directory where all shared drives are under VMPlayer, 'Meow' is the name of my\n",
    "# shared drive,'WGS/references' is the folder directory under the shared drive, and 'ShigellaRef6.fasta' is the\n",
    "# name of the current reference sequence database. \n",
    "# Remove the current line beginning with \"ShigellaRef\" dictating where the default reference sequence database is.\n",
    "\n",
    "# import os\n",
    "# AbsRef_dir = '/mnt/hgfs/Meow/WGS/references/ShigellaRef6.fasta'\n",
    "# ShigellaRef = os.path.relpath(AbsRef_dir, os.getcwd())\n",
    "\n",
    "# relative directory of the reference sequence database\n",
    "ShigellaRef = \"../../references/ShigellaRef6.fasta\"\n",
    "if os.path.isfile(ShigellaRef) == False:\n",
    "    print(\"Error: reference sequence database does not exist!\")\n",
    "    exit()\n",
    "\n",
    "# generate a index file for reference\n",
    "dir_path = os.path.dirname(os.path.realpath(ShigellaRef))\n",
    "rel_dir = os.path.relpath(dir_path, os.getcwd())\n",
    "Refname = ShigellaRef[ShigellaRef.rfind('/')+1:ShigellaRef.rfind('.')]\n",
    "Refname = Refname+\".mmi\"\n",
    "mmi_index = os.path.join(rel_dir, Refname)\n",
    "if os.path.isfile(mmi_index) == False:\n",
    "    print(\"building Reference sequence index.......\")\n",
    "    !minimap2 -d $mmi_index $ShigellaRef\n",
    "# another index needed to generate\n",
    "fai_index = ShigellaRef+'.fai'\n",
    "if os.path.isfile(fai_index) == False:\n",
    "    !samtools faidx $ShigellaRef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import datetime\n",
    "now = datetime.datetime.now().strftime(\"%Y%m%d\")\n",
    "import papermill as pm\n",
    "outputs = []\n",
    "for Sample in samples:\n",
    "    reads = []\n",
    "    for file in files:\n",
    "        if file.startswith(Sample+\"_\"):\n",
    "            reads.append(file)\n",
    "    if len(reads) == 2:\n",
    "        output = Sample + \"_\" + str(threshold) + \"_\" + now + \".ipynb\"\n",
    "        outputs.append(output); print(\" \")\n",
    "        pm.execute_notebook('batch_010819.ipynb', output,\n",
    "                            dict(Sample=Sample, read1 = reads[0], read2 = reads[1], \n",
    "                                 ShigellaRef = ShigellaRef, fastp = fastp, threshold = threshold)) \n",
    "        !jupyter nbconvert --to html $output\n",
    "    else: print('\\n'+Sample + \" does not have pair-end reads! Please re-examine.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if there are fastq files also process them\n",
    "files = glob.glob(\"*.fastq\")\n",
    "names = []\n",
    "for file in files:\n",
    "    names.append(file[:file.find(\"_\")])\n",
    "fastqsamples = list(set(names))\n",
    "fastqsamples.sort()\n",
    "samples = samples + fastqsamples\n",
    "for Sample in fastqsamples:\n",
    "    size = 0\n",
    "    for file in files:\n",
    "        if file.startswith(Sample+\"_\"):\n",
    "            size += os.path.getsize(file)\n",
    "    sizes.append(round(size/1024000, 1))\n",
    "\n",
    "for Sample in fastqsamples:\n",
    "    reads = []\n",
    "    for file in files:\n",
    "        if file.startswith(Sample+\"_\"):\n",
    "            reads.append(file)\n",
    "    if len(reads) == 2:\n",
    "        output = Sample + \"_\" + str(threshold) + \"_\" + now + \".ipynb\"\n",
    "        outputs.append(output); print(\" \")\n",
    "        pm.execute_notebook('batch_010819.ipynb', output,\n",
    "                            dict(Sample=Sample, read1 = reads[0], read2 = reads[1], \n",
    "                                 ShigellaRef = ShigellaRef, fastp = fastp, threshold = threshold)) \n",
    "        !jupyter nbconvert --to html $output\n",
    "    else: print('\\n'+Sample + \" does not have pair-end reads! Please re-examine.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Date of analysis:\", now)\n",
    "\n",
    "try: samples\n",
    "except: samples = []\n",
    "    \n",
    "if len(samples) > 0:\n",
    "    sizetable = pd.DataFrame({'Sample': samples, 'Size (MB)': sizes})\n",
    "\n",
    "    print(\"Threshold level for gene coverage: \", threshold, \"%\")\n",
    "    print(len(outputs), \" samples were analyzed: \\n\")\n",
    "    samples = []\n",
    "    predictions = []\n",
    "    pINV = []\n",
    "    ShigaToxin=[]\n",
    "    Enterotoxin=[]\n",
    "\n",
    "    for output in outputs:\n",
    "        nb = pm.read_notebook(output)\n",
    "        sampleIndex = nb.dataframe.index[nb.dataframe['name']==\"Sample\"].tolist()[0]\n",
    "        predictionIndex = nb.dataframe.index[nb.dataframe['name']==\"prediction\"].tolist()[0]\n",
    "        pINVindex = nb.dataframe.index[nb.dataframe['name']==\"pINV\"].tolist()[0]\n",
    "        stxIndex = nb.dataframe.index[nb.dataframe['name']==\"ShigaToxin\"].tolist()[0]\n",
    "        enteroIndex = nb.dataframe.index[nb.dataframe['name']=='Enterotoxin'].tolist()[0]\n",
    "        samples.append(nb.dataframe.iloc[sampleIndex]['value'])\n",
    "        predictions.append(nb.dataframe.iloc[predictionIndex]['value'])\n",
    "        pINV.append(nb.dataframe.iloc[pINVindex]['value'])\n",
    "        ShigaToxin.append(nb.dataframe.iloc[stxIndex]['value'])\n",
    "        Enterotoxin.append(nb.dataframe.iloc[enteroIndex]['value'])\n",
    "        os.remove(output) # delete notebook to keep directory tidy\n",
    "  \n",
    "    table = pd.DataFrame({'Sample': samples, 'Serotype prediction': predictions,'Invasion plasmid': pINV,\n",
    "                          'Shiga Toxin': ShigaToxin, 'Enterotoxin': Enterotoxin})\n",
    "    final = pd.merge(left=sizetable, right = table, left_on=\"Sample\", right_on = \"Sample\", how = 'left')\n",
    "    final = final.sort_values(by=['Sample'])\n",
    "    from IPython.display import display, HTML\n",
    "    display(HTML(final.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Process assembled genomes in fasta format\n",
    "files = glob.glob(\"*.fasta\")\n",
    "names = []\n",
    "for file in files:\n",
    "    names.append(file[:file.rfind(\".fasta\")])\n",
    "fastasamples = list(set(names))\n",
    "fastasamples.sort()\n",
    "\n",
    "fastasizes = []\n",
    "for Sample in fastasamples:\n",
    "    size = os.path.getsize(file)\n",
    "    fastasizes.append(round(size/1024000, 1))\n",
    "\n",
    "fastaoutputs =[]\n",
    "for Sample in fastasamples:\n",
    "    output = Sample + \"_\" + str(threshold) + \"_\" + now + \".ipynb\"\n",
    "    fastaoutputs.append(output); print(\" \")\n",
    "    pm.execute_notebook('batch_010819.ipynb', output,\n",
    "                        dict(Sample=Sample, read1 = Sample+'.fasta', \n",
    "                                 ShigellaRef = ShigellaRef, fastp = fastp, threshold = threshold)) \n",
    "    !jupyter nbconvert --to html $output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary of serotype prediction results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Display final table for assembled contigs\n",
    "try: fastasamples\n",
    "except: fastasamples = []\n",
    "\n",
    "if len(fastasamples) > 0:\n",
    "    fastasize = pd.DataFrame({'Sample': fastasamples, 'Size (MB)': fastasizes})\n",
    "\n",
    "    print(\"Threshold level for gene coverage: \", threshold, \"%\")\n",
    "    print(len(fastaoutputs), \" samples were analyzed: \\n\")\n",
    "    samples = []\n",
    "    predictions = []\n",
    "    pINV = []\n",
    "    ShigaToxin=[]\n",
    "    Enterotoxin=[]\n",
    "\n",
    "    for output in fastaoutputs:\n",
    "        nb = pm.read_notebook(output)\n",
    "        sampleIndex = nb.dataframe.index[nb.dataframe['name']==\"Sample\"].tolist()[0]\n",
    "        predictionIndex = nb.dataframe.index[nb.dataframe['name']==\"prediction\"].tolist()[0]\n",
    "        pINVindex = nb.dataframe.index[nb.dataframe['name']==\"pINV\"].tolist()[0]\n",
    "        stxIndex = nb.dataframe.index[nb.dataframe['name']==\"ShigaToxin\"].tolist()[0]\n",
    "        enteroIndex = nb.dataframe.index[nb.dataframe['name']=='Enterotoxin'].tolist()[0]\n",
    "        samples.append(nb.dataframe.iloc[sampleIndex]['value'])\n",
    "        predictions.append(nb.dataframe.iloc[predictionIndex]['value'])\n",
    "        pINV.append(nb.dataframe.iloc[pINVindex]['value'])\n",
    "        ShigaToxin.append(nb.dataframe.iloc[stxIndex]['value'])\n",
    "        Enterotoxin.append(nb.dataframe.iloc[enteroIndex]['value'])\n",
    "        os.remove(output) # delete notebook to keep directory tidy\n",
    "  \n",
    "    fastatable = pd.DataFrame({'Sample': samples, 'Serotype prediction': predictions,\n",
    "                               'Invasion plasmid': pINV, 'Shiga Toxin': ShigaToxin, 'Enterotoxin': Enterotoxin})\n",
    "    fastafinal = pd.merge(left=fastasize, right = fastatable, left_on=\"Sample\", right_on = \"Sample\", how = 'left')\n",
    "    fastafinal = fastafinal.sort_values(by=['Sample'])\n",
    "    from IPython.display import display, HTML\n",
    "    display(HTML(fastafinal.to_html(index=False)))\n",
    "\n",
    "\n",
    "#print(nb.dataframe.iloc[3, 1], '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "The raw code for this notebook is by default hidden for easier reading.\n",
    "To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>.''')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
